File: mcp-client-bridge/src/reconnection-manager.ts
/**
 * Manages auto-reconnection to the Orchestrator with exponential backoff.
 * Supports multi-URL failover via UrlManager.
 */

import { BridgeConfig } from './bridge-config.js';
import { HttpStreamableClient } from './http-streamable-client.js';
import { UrlManager } from './url-manager.js';

export enum BridgeState {
  DISCONNECTED = 'DISCONNECTED',
  CONNECTING = 'CONNECTING',
  CONNECTED = 'CONNECTED',
  RECONNECTING = 'RECONNECTING',
}

export class ReconnectionManager {
  private attempt = 0;
  state: BridgeState = BridgeState.DISCONNECTED;
  private readonly urlManager: UrlManager;

  constructor(
    private readonly config: BridgeConfig,
    private readonly client: HttpStreamableClient,
  ) {
    this.urlManager = new UrlManager(config.orchestratorUrls);
  }

  /** Attempt initial connection trying all URLs sequentially. */
  async connectWithRetry(): Promise<boolean> {
    this.state = BridgeState.CONNECTING;
    this.urlManager.clearErrors();

    for (let i = 0; i < this.urlManager.urlCount; i++) {
      const url = this.urlManager.activeUrl;
      const idx = this.urlManager.urlIndex;
      console.error(
        `[mcp-bridge] Trying URL ${idx + 1}/${this.urlManager.urlCount}: ${url}`,
      );

      try {
        if (await this.client.initialize(url)) {
          this.state = BridgeState.CONNECTED;
          this.attempt = 0;
          return true;
        }
      } catch (err: unknown) {
        const msg = err instanceof Error ? err.message : String(err);
        this.urlManager.markFailed(url, msg);
        console.error(
          `[mcp-bridge] URL ${idx + 1}/${this.urlManager.urlCount} failed: ${msg}`,
        );
      }

      if (this.urlManager.hasNext()) {
        this.urlManager.advance();
      }
    }

    this.reportErrors();
    this.state = BridgeState.DISCONNECTED;
    return false;
  }

  /** Background reconnection with retry + URL rotation. */
  async reconnectLoop(): Promise<void> {
    if (!this.config.reconnectEnabled) return;
    this.state = BridgeState.RECONNECTING;

    // Phase 1: Retry active URL
    for (let i = 0; i < this.config.maxRetryBeforeRotate; i++) {
      const delay = this.calculateBackoff(i);
      console.error(
        `[mcp-bridge] Retry ${i + 1}/${this.config.maxRetryBeforeRotate} ` +
        `for ${this.urlManager.activeUrl} in ${delay}ms`,
      );
      await this.sleep(delay);

      this.client.resetSession();
      if (await this.client.initialize(this.urlManager.activeUrl)) {
        this.state = BridgeState.CONNECTED;
        this.attempt = 0;
        console.error('[mcp-bridge] Reconnected successfully');
        return;
      }
    }

    // Phase 2: Rotate to other URLs
    if (this.urlManager.urlCount > 1) {
      this.urlManager.clearErrors();
      this.urlManager.markFailed(this.urlManager.activeUrl, 'Exhausted retries');

      while (this.urlManager.hasNext()) {
        const nextUrl = this.urlManager.advance();
        console.error(
          `[mcp-bridge] Switching to URL ` +
          `${this.urlManager.urlIndex + 1}/${this.urlManager.urlCount}: ${nextUrl}`,
        );

        this.client.resetSession();
        if (await this.client.initialize(nextUrl)) {
          this.state = BridgeState.CONNECTED;
          this.attempt = 0;
          return;
        }
        this.urlManager.markFailed(nextUrl, 'Connection failed');
      }

      this.reportErrors();
      this.urlManager.reset();
    }

    // Phase 3: Infinite backoff loop on first URL
    await this.infiniteReconnect();
  }

  private async infiniteReconnect(): Promise<void> {
    while (!this.client.isConnected) {
      const delay = this.calculateBackoff(this.attempt);
      console.error(
        `[mcp-bridge] Reconnecting in ${delay}ms (attempt ${this.attempt})`,
      );
      await this.sleep(delay);

      this.client.resetSession();
      if (await this.client.initialize(this.urlManager.activeUrl)) {
        this.state = BridgeState.CONNECTED;
        this.attempt = 0;
        console.error('[mcp-bridge] Reconnected successfully');
        return;
      }
      this.attempt++;
    }
  }

  private reportErrors(): void {
    const errors = this.urlManager.getErrors();
    const lines = errors.map(e => `  - ${e.url}: ${e.error}`);
    console.error(`[mcp-bridge] All URLs failed:\n${lines.join('\n')}`);
  }

  private calculateBackoff(attempt: number): number {
    const delay = this.config.baseReconnectDelayMs * Math.pow(2, attempt);
    return Math.min(delay, this.config.maxReconnectDelayMs);
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

File: orchestrator-bridge/src/main/kotlin/com/orchestrator/mcp/bridge/ReconnectionManager.kt
package com.orchestrator.mcp.bridge

import com.orchestrator.mcp.core.util.RetryUtils
import kotlinx.coroutines.delay
import org.slf4j.LoggerFactory

/**
 * Manages auto-reconnection with multi-URL failover and exponential backoff.
 * Max delay capped at 15 seconds per TDD spec.
 */
class ReconnectionManager(
    private val config: BridgeConfig,
    private val client: HttpStreamableClient
) {
    private val logger = LoggerFactory.getLogger(ReconnectionManager::class.java)
    private val urlManager = UrlManager(config.orchestratorUrls)
    private var attempt = 0
    var state: BridgeState = BridgeState.DISCONNECTED

    /** Try all URLs sequentially for initial connection. */
    suspend fun connectWithRetry(): Boolean {
        state = BridgeState.CONNECTING
        urlManager.clearErrors()

        repeat(urlManager.urlCount) {
            val url = urlManager.activeUrl
            val idx = urlManager.urlIndex
            logger.info("Trying URL ${idx + 1}/${urlManager.urlCount}: $url")

            try {
                if (client.initialize(url, config.connectionTimeoutMs)) {
                    state = BridgeState.CONNECTED
                    attempt = 0
                    return true
                }
            } catch (e: Exception) {
                urlManager.markFailed(url, e.message ?: "Unknown error")
                logger.warn("URL ${idx + 1}/${urlManager.urlCount} failed: ${e.message}")
            }

            if (urlManager.hasNext()) urlManager.advance()
        }

        reportErrors()
        state = BridgeState.DISCONNECTED
        return false
    }

    /** Reconnect with retry on active URL, then rotate. */
    suspend fun reconnectLoop() {
        if (!config.reconnectEnabled) return
        state = BridgeState.RECONNECTING

        // Phase 1: Retry active URL
        repeat(config.maxRetryBeforeRotate) { i ->
            val delayMs = RetryUtils.calculateBackoff(
                i, config.baseReconnectDelayMs, config.maxReconnectDelayMs
            )
            logger.warn(
                "Retry ${i + 1}/${config.maxRetryBeforeRotate} " +
                "for ${urlManager.activeUrl} in ${delayMs}ms"
            )
            delay(delayMs)
            client.resetSession()
            if (client.initialize(urlManager.activeUrl, config.connectionTimeoutMs)) {
                state = BridgeState.CONNECTED
                attempt = 0
                logger.info("Reconnected successfully")
                return
            }
        }

        // Phase 2: Rotate to other URLs
        if (urlManager.urlCount > 1) {
            urlManager.clearErrors()
            urlManager.markFailed(urlManager.activeUrl, "Exhausted retries")

            while (urlManager.hasNext()) {
                val nextUrl = urlManager.advance()
                logger.warn(
                    "Switching to URL ${urlManager.urlIndex + 1}" +
                    "/${urlManager.urlCount}: $nextUrl"
                )
                client.resetSession()
                if (client.initialize(nextUrl, config.connectionTimeoutMs)) {
                    state = BridgeState.CONNECTED
                    attempt = 0
                    return
                }
                urlManager.markFailed(nextUrl, "Connection failed")
            }

            reportErrors()
            urlManager.reset()
        }

        // Phase 3: Infinite backoff loop
        infiniteReconnect()
    }

    private suspend fun infiniteReconnect() {
        while (!client.isConnected) {
            val delayMs = RetryUtils.calculateBackoff(
                attempt, config.baseReconnectDelayMs, config.maxReconnectDelayMs
            )
            logger.info("Reconnecting in ${delayMs}ms (attempt $attempt)")
            delay(delayMs)
            client.resetSession()
            if (client.initialize(urlManager.activeUrl, config.connectionTimeoutMs)) {
                state = BridgeState.CONNECTED
                attempt = 0
                logger.info("Reconnected successfully")
                return
            }
            attempt++
        }
    }

    private fun reportErrors() {
        val errors = urlManager.getErrors()
        val report = errors.joinToString("\n") { "  - ${it.url}: ${it.error}" }
        logger.error("All URLs failed:\n$report")
    }
}

enum class BridgeState {
    DISCONNECTED,
    CONNECTING,
    CONNECTED,
    RECONNECTING
}
